specs/DesignDiscussions/InputValidation.md

```diff
@@ -0,0 +1,260 @@
+# Input Validation
+
+Input validation was originally spec'd in issue [#3081](https://github.com/microsoft/AdaptiveCards/issues/3081). This document covers the basics from that spec, and continues the discussion of additional concerns that it raised.
+
+## Schema Additions to Inputs for Validation
+
+``` json
+{
+    "type": "Input.Text",
+    "id": "username",
+    "isRequired": true,
+    "regex": "^\\w{5,50}$",
+    "errorMessage": "Username must be between 5 and 50 characters long."
+}
+```
+The main additions to the schema for input validation are the `"isRequired"`, `"regex"`, and `"errorMessage"` properties on inputs. These properties allow a card author to specify whether or not an input is required, what regular expression defines valid input, and what error message should be shown in case of invalid user input.
+
+Although `regex` would only be supported on `Input.Text`, the `isRequired` and `errorMessage` properties would be supported on all `Input.*` types.
+
+## Schema Additions to Adaptive Card for Validation
+
+In order to allow the card author to visually indicate whether inputs in a card are required or optional, we will add a new property, `showRequiredInputHints`. When this is true, visual indicators will be added to required inputs inputs to signal to the user that they are required. This property will default to true.
+
+```json
+{
+	"type": "AdaptiveCard",
+	"version": "1.3",
+	"showRequiredInputHints":true,
+	"body": [
+		{
+			"type": "TextBlock",
+			"text": "Please enter your name:"
+		},
+		{
+			"type": "Input.Text",
+			"id": "nameInput",
+			"isRequired": true
+		}
+	],
+	"$schema": "http://adaptivecards.io/schemas/adaptive-card.json"
+}
+```
+
+> Note: We may also want to add a `showOptionalInputHints` to allow card authors to instead (or additionally) mark optional inputs visually. Because the most likely design for this would be to use the word "optional," and because that would require an all up localization story we don't yet have, this is out of scope for v1 of this feature.
+
+## Validation Rendering
+
+If showRequiredInputHints is true, Required inputs should be marked with a *. In cases where a label has been specified ([Issue #203](https://github.com/microsoft/AdaptiveCards/issues/203)), the required indicator should be placed next to the input's label.
+
+![img](assets/InputValidation/Required.PNG)
+
+The error text and invalid indication will be rendered in the host's `attention` color.
+
+![img](assets/InputValidation/InputValidationExample.png)
+
+## Input Validation Behavior
+
+### When to Validate
+
+Issue [#3081](https://github.com/microsoft/AdaptiveCards/issues/3081) covers research and best practices behind when to validate. The conclusion reached there is as follows:
```

Comment:
Lots of discussion on this. Outlook.com has behavior that allows clicking the button and *then* validates on navigate.

Choice sets and other inputs cause problems with the model spec'd here.

Matt feels strongly we should scroll the first bad one into view.

I want to follow up with WinUI to see if they've thought about this.

Reply:
Updated the spec to indicate we should scroll the first invalid input into view. Label discussion will continue in the label spec.

---
In reply to: [355034372](https://github.com/microsoft/AdaptiveCards/pull/3663#discussion_r355034372) [](ancestors = 355034372)